File: backend/rag_agent.py
from typing import Dict, List, Optional
from azure.search.documents import SearchClient
from azure.core.credentials import AzureKeyCredential
from openai import AzureOpenAI
import numpy as np
from .model_types import Coordinates4D, SearchDocument
from .advanced_ai_engine import AdvancedAIEngine

class RAGAgent:
    def __init__(self, config: Dict[str, str]):
        self.search_client = SearchClient(
            endpoint=config["search_endpoint"],
            index_name="documents",
            credential=AzureKeyCredential(config["search_key"])
        )
        
        self.openai = AzureOpenAI(
            api_key=config["openai_key"],
            api_version="2024-02-15-preview",
            azure_endpoint=config["openai_endpoint"]
        )
        
        self.deployment_name = config["deployment_name"]
        
        # Initialize Advanced AI Engine
        self.advanced_engine = AdvancedAIEngine(config)

    async def process_query(self, query: str, expertise_level: int) -> Dict:
        # Get embeddings and basic search results
        embedding_response = await self.openai.embeddings.create(
            model="text-embedding-3-small",
            input=query
        )
        query_embedding = embedding_response.data[0].embedding

        # Get advanced analysis
        advanced_results = await self.advanced_engine.analyze(
            query=query,
            expertise_level=expertise_level,
            embeddings=query_embedding
        )

        # Combine with semantic search
        search_results = await self.search_client.search(
            search_text=query,
            select=["id", "content", "metadata", "coordinates", "_score"],
            query_type="semantic",
            semantic_configuration_name="default",
            top=5
        )

        documents = [doc for doc in search_results]

        # Generate visualization data with advanced insights
        visualization_data = self._generate_visualization(
            documents, 
            query_embedding,
            advanced_results.get('insights', {})
        )

        return {
            "query": query,
            "semantic_results": [{
                "id": doc["id"],
                "content": doc["content"],
                "coordinates": doc["coordinates"],
                "relevance_score": doc.get("_score", 0)
            } for doc in documents],
            "visualization_data": visualization_data,
            "explanation_tree": advanced_results.get('explanation_tree', {}),
            "response": advanced_results.get('response', '')
        }

    def _construct_prompt(self, query: str, documents: List[Dict]) -> str:
        return f"""
        Query: {query}

        Context:
        {chr(10).join(doc["content"] for doc in documents)}

        Please provide a detailed response based on the above context.
        """

    def _generate_visualization(self, documents: List[Dict], query_embedding: List[float], advanced_insights: Dict) -> Dict:
        nodes = [{
            "id": doc["id"],
            "coordinates": doc["coordinates"],
            "category": doc["metadata"].get("category", "unknown"),
            "relevance": self._calculate_relevance(
                doc.get("embedding", []), 
                query_embedding
            )
        } for doc in documents]

        edges = self._generate_edges(nodes)
        heatmap = self._generate_heatmap(documents)

        return {
            "space_mapping": {"nodes": nodes, "edges": edges},
            "heatmap_data": heatmap,
            "advanced_insights": advanced_insights
        }

    def _calculate_relevance(self, v1: List[float], v2: List[float]) -> float:
        if not v1 or not v2:
            return 0.0
        return np.dot(v1, v2) / (np.linalg.norm(v1) * np.linalg.norm(v2))

    def _generate_edges(self, nodes: List[Dict]) -> List[Dict]:
        edges = []
        for i, node in enumerate(nodes):
            for other in nodes[i+1:]:
                edges.append({
                    "source": node["id"],
                    "target": other["id"],
                    "weight": self._calculate_similarity(
                        node["coordinates"],
                        other["coordinates"]
                    )
                })
        return edges

    def _generate_heatmap(self, documents: List[Dict]) -> Dict:
        categories = list(set(doc["metadata"].get("category") for doc in documents))
        requirements = list(set(doc["metadata"].get("requirement") for doc in documents))
        
        matrix = []
        for req in requirements:
            row = []
            for cat in categories:
                relevant_docs = [
                    doc for doc in documents 
                    if doc["metadata"].get("category") == cat and 
                       doc["metadata"].get("requirement") == req
                ]
                row.append(max((doc.get("_score", 0) for doc in relevant_docs), default=0))
            matrix.append(row)
            
        return {
            "matrix": matrix,
            "categories": categories,
            "requirements": requirements
        }

    def _generate_explanation_tree(self, query: str, documents: List[Dict]) -> Dict:
        return {
            "step": "Query Analysis",
            "reasoning": f"Analyzing query: {query}",
            "confidence": 0.9,
            "evidence": [{
                "content": doc["content"][:200],
                "source": doc["id"],
                "relevance": doc.get("_score", 0)
            } for doc in documents[:3]],
            "subSteps": []
        }

File: src/lib/rag_agent.ts
import { SearchClient, AzureKeyCredential, ExtractiveQueryAnswer, ExtractiveQueryCaption, QueryType, SearchOptions, SemanticSearchOptions } from "@azure/search-documents";
import { AzureOpenAI } from "openai";
import { ExplanationNode } from '../types/explanation';
import { Coordinates4D } from '../types/shared';

interface Persona {
  type: 'legal' | 'financial' | 'compliance';
  confidence: number;
  analysis: string;
}

interface RAGResponse {
  query: string;
  semantic_results: Array<{
    id: string;
    content: string;
    coordinates: Coordinates4D;
    relevance_score: number;
  }>;
  personas: Persona[];
  visualization_data: {
    space_mapping: {
      nodes: Array<{
        id: string;
        coordinates: Coordinates4D;
        category: string;
        relevance: number;
      }>;
      edges: Array<{
        source: string;
        target: string;
        weight: number;
      }>;
    };
    heatmap_data: {
      matrix: number[][];
      categories: string[];
      requirements: string[];
    };
  };
  explanation_tree: ExplanationNode;
  response: string;
}

interface SearchDocument {
  id: string;
  content: string;
  metadata: Record<string, any>;
  coordinates: Coordinates4D;
  embedding?: number[];
  _score?: number;
}

export class RAGAgent {
  private searchClient: SearchClient<SearchDocument>;
  private openai: AzureOpenAI;

  constructor(config: {
    azure_endpoint: string;
    azure_deployment: string;
    search_endpoint: string;
    search_key: string;
  }) {
    this.searchClient = new SearchClient(
      config.search_endpoint,
      "documents",
      new AzureKeyCredential(config.search_key)
    );

    this.openai = new AzureOpenAI({
      apiKey: config.azure_deployment,
      endpoint: config.azure_endpoint,
      apiVersion: "2024-08-01-preview"
    });
  }

  async process_query(query: string, context: Record<string, any>): Promise<RAGResponse> {
    // Get embeddings
    const embedding = await this.openai.embeddings.create({
      model: "text-embedding-3-small",
      input: query
    });

    // Search in 4D space
    const searchOptions: SearchOptions<SearchDocument> = {
      queryType: 'semantic',
      semanticSearchOptions: {
        queryFields: ['content'],
        prioritizedFields: {
          titleFields: [],
          keywordsFields: [],
          contentFields: ['content']
        }
      },
      select: ['id', 'content', 'metadata', 'coordinates', '_score'],
      top: 5,
      includeTotalCount: true,
      queryLanguage: 'en-us',
      semanticConfiguration: 'default',
      searchFields: ['content'],
      orderBy: ['_score desc']
    };

    const searchResults = await this.searchClient.search("*", searchOptions);

    // Process results
    const documents = [];
    for await (const result of searchResults.results) {
      documents.push(result.document);
    }

    // Generate response
    const completion = await this.openai.chat.completions.create({
      model: "gpt-4",
      messages: [
        { role: "system", content: "You are a helpful assistant." },
        { role: "user", content: this.constructPrompt(query, documents) }
      ]
    });

    // Generate 4D visualization
    const visualizationData = await this.generate4DVisualization(documents, embedding.data[0].embedding);

    return {
      query,
      semantic_results: documents.map(doc => ({
        id: doc.id,
        content: doc.content,
        coordinates: doc.coordinates,
        relevance_score: doc._score || 0
      })),
      personas: [],
      visualization_data: visualizationData,
      explanation_tree: {
        step: "Query Analysis",
        reasoning: `Analyzing query: ${query}`,
        confidence: 0.9,
        evidence: documents.slice(0, 3).map(doc => ({
          content: doc.content,
          source: doc.id,
          relevance: doc._score || 0
        })),
        subSteps: []
      } as ExplanationNode,
      response: completion.choices[0]?.message?.content || ""
    };
  }

  private constructPrompt(query: string, documents: any[]): string {
    return `
    Query: ${query}

    Context:
    ${documents.map(doc => doc.content).join('\n\n')}

    Please provide a detailed response based on the above context.
    `;
  }

  private async generate4DVisualization(documents: any[], query_embedding: number[]): Promise<RAGResponse['visualization_data']> {
    // Generate network visualization
    const nodes = documents.map(doc => ({
      id: doc.id,
      coordinates: doc.coordinates,
      category: doc.metadata.category,
      relevance: this.calculateRelevance(doc.embedding, query_embedding)
    }));

    const edges = this.generateEdges(nodes);
    
    // Generate heatmap
    const heatmap = this.generateComplianceHeatmap(documents);

    return {
      space_mapping: { nodes, edges },
      heatmap_data: heatmap
    };
  }

  private generateEdges(nodes: any[]) {
    return nodes.flatMap((node, i) => 
      nodes.slice(i + 1).map(other => ({
        source: node.id,
        target: other.id,
        weight: this.calculateSimilarity(node.coordinates, other.coordinates)
      }))
    );
  }

  private generateComplianceHeatmap(documents: any[]) {
    const categorySet = new Set(documents.map(d => d.metadata.category));
    const requirementSet = new Set(documents.map(d => d.metadata.requirement));
    
    const categories = Array.from(categorySet);
    const requirements = Array.from(requirementSet);
    
    const matrix = requirements.map(req =>
      categories.map(cat => {
        const relevantDocs = documents.filter(d => 
          d.metadata.category === cat && 
          d.metadata.requirement === req
        );
        return relevantDocs.length ? Math.max(...relevantDocs.map(d => d.score)) : 0;
      })
    );

    return { matrix, categories, requirements };
  }

  private calculateRelevance(v1: number[], v2: number[]): number {
    // Cosine similarity
    const dotProduct = v1.reduce((sum, a, i) => sum + a * v2[i], 0);
    const mag1 = Math.sqrt(v1.reduce((sum, a) => sum + a * a, 0));
    const mag2 = Math.sqrt(v2.reduce((sum, a) => sum + a * a, 0));
    return dotProduct / (mag1 * mag2);
  }

  private calculateSimilarity(c1: Coordinates4D, c2: Coordinates4D): number {
    const dims = ['x', 'y', 'z', 'e'] as const;
    const sqSum = dims.reduce((sum, dim) => sum + Math.pow(c1[dim] - c2[dim], 2), 0);
    return 1 / (1 + Math.sqrt(sqSum));
  }
}
